### ISSUE TITLE
feat(native): expose opt-in token-indexed decode traces

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#9071]
## Parent context

Prerequisite for #8385. Related consumer #8775 measures user-visible interactivity; this leaf owns only fak-native token-position evidence.

## Problem frame

Centrality: Enabling — long Qwen3.8 generations cannot be promotion-gated by position while fak exposes only aggregate decode time.

- P1 managed context: advanced — the evidence names output position as recurrent/KV state grows.
- P2 net-true efficiency: advanced — timing begins at the actual native decode commit seam, not at arbitrary SSE fragment boundaries.
- P3 bounded adaptation: advanced — tracing is request opt-in, schema-versioned, and absent on the default path.
- P4 integrated operations: advanced — benchmark callers receive a fail-closed receipt instead of inferring token counts from content fragments.

For: operators qualifying fak-native Qwen3.8 on Apple Silicon.
Problem: aggregate tok/s can hide position-dependent degradation, but SSE chunks are not guaranteed to equal tokens.
Today: `InKernelPlanner.CompleteStream` buffers the whole native completion and emits it as one content fragment; proxy captures can carry `token_ids:null`.
Better because: a native-only response trace records a monotonic timestamp at each generated-token commit and refuses unsupported proxy/stream requests.
Witness: a deterministic synthetic-model request returns exact 1-based token indices and monotonic elapsed times, while default, proxy, and ambiguous streaming paths emit no trace.

## Current evidence

- `internal/agent/inkernel_planner.go` currently implements native `CompleteStream` by calling buffered `Complete` and sending the whole answer once.
- `internal/agent/inkernel_decode.go` `decodeLane.decodeOne` is the shared serial/batched token-commit seam: a token is counted only after the stop check and emit.
- `internal/gateway/wire.go` streamed deltas carry content but no token index.
- Captured OpenAI-compatible fixtures under `cmd/streamcapture/testdata/captures/` include `token_ids:null`; content-fragment count is therefore not a token-count witness.
- The motivating llama.cpp report #27444 was closed after the reporter identified a CUDA-header/runtime ABI mismatch, not a llama.cpp Qwen kernel defect. The measurement need remains useful, but this issue makes no causal claim about llama.cpp: https://github.com/ggml-org/llama.cpp/issues/27444#issuecomment-5374360947

## Working spine

1. Add an explicit non-standard request opt-in, `fak_decode_trace:true`, accepted only for non-streaming fak-native completions.
2. Thread a decode-trace option into the in-kernel planner. Default-off execution performs no per-token timestamp or trace allocation.
3. At `decodeLane.decodeOne`, record `{token_index, elapsed_ns}` after a non-stop token is emitted/counted. Indices are 1-based and consecutive; elapsed time uses Go's monotonic clock relative to native decode start. Clear first-attempt events before an OOM retry.
4. Carry the trace through `agent.Completion` and expose it under the response `fak` extension with a versioned schema and explicit engine `fak-native`.
5. Reject `fak_decode_trace:true` before inference when the selected planner is proxy/non-native or the request also sets `stream:true`. Never synthesize traces from SSE chunks.

## In scope

- Native, non-streaming, request-opt-in token-index/timestamp trace.
- Exact engine/schema provenance and monotonic/consecutive validation.
- Fail-closed unsupported-path behavior.
- Synthetic deterministic tests for serial and opt-in batched decode.
- Default-off allocation/shape regression witness.

## Out of scope

- Window/slope/confidence aggregation; #8385 consumes the raw trace.
- A general profiler, token text/IDs/logits, tool pauses, or provider interactivity scoring.
- Changing model math, sampling, default streaming behavior, or gateway compatibility.
- Claiming that the CUDA report reproduces on Apple Silicon.

## Done condition

One opt-in non-streaming fak-native completion returns a schema-versioned engine-authored trace whose event count and consecutive indices equal generated completion tokens, while default requests stay inert and proxy, streamed, malformed, or retry-duplicated traces refuse before inference.

## Witness

`go test ./internal/agent ./internal/gateway -run \u0027Test(InKernelDecodeTrace|ChatDecodeTrace)\u0027 -count=1`

This single focused witness binds exact native indices, monotonic fake-clock timing, OOM-retry reset, default-off inertness, proxy/stream pre-inference refusal, and serial/batched token parity.

## Likely files

- `internal/agent/chat.go`
- `internal/agent/sample_options.go`
- `internal/agent/inkernel_planner.go`
- `internal/agent/inkernel_decode.go`
- `internal/agent/inkernel_decode_test.go`
- `internal/gateway/wire.go`
- `internal/gateway/http.go`
- `internal/gateway/gateway_test.go`

## Coordination

Lane: agent+gateway. No GPU required. Do not dispatch concurrently with another worker editing the listed agent/gateway decode paths. #8385 remains blocked on this trace and owns the later Mac campaign; #8775 may consume the same trace but must not broaden this leaf.

## Batch policy

Standalone. Dispatch only after DOS reports the agent+gateway tree free and Codex preflight returns `SPAWN_OK`.

## Work unit

leaf

## Expected steps

5

## Completion standard

development

## Category

instrumentation

## Layer

qwen38-native-decode-trace

## Outcome

A non-streaming fak-native completion can return an opt-in, versioned, token-indexed monotonic decode trace; every unsupported path refuses before inference.

## Current state

Native generation has one exact token seam in decodeLane.decodeOne, but only whole-loop timing survives. Native CompleteStream emits the completed answer once, and gateway/proxy fragment shapes carry no trustworthy token index.

## Why this is next

#8385 cannot compute honest output-position windows until the native engine authors a raw token trace. This leaf is the smallest source-of-truth seam and requires no GPU.

## Acceptance gate

The focused agent+gateway witness passes at committed tip and proves exact indices, monotonic timestamps, retry reset, default-off inertness, serial/batched token parity, and zero model calls on proxy or stream refusal. The response receipt must name schema fak.native-decode-trace/1 and engine fak-native.

## Closure binding

One Conventional-Commits commit closes #9071, ends with a recognized (fak agent) trailer, touches only the listed paths, and is independently bound with dos verify after the focused tests pass.

## Lane

agent

## Trigger

#8385 needs position-window evidence, while current native and proxy streaming contracts expose content fragments rather than tokenizer events.

## Placement

Generation: gen/now. Priority: P1. Milestone: Generation G0 - Now / Immediate. Lane: agent. Work class: class:dev.

## Work scale

S1

## Model tier

- Required model tier: tier/T2-required
- Optimal model tier: tier/T1-optimal

## Assumptions

- decodeLane.decodeOne remains the shared serial and batched native token-commit seam.
- A request-scoped optional capability can reject proxy and streamed trace requests before inference.
- A monotonic injectable clock can witness ordering without changing model math.

## Confusion risks

- One SSE chunk is not one token.
- Token timestamping must occur after token-ID stop filtering and before detokenization; a string-suffix stop still counts the generated token.
- A trace from a proxy or comparator is not a fak-native receipt.
- Default-off tracing must not allocate or call time.Now per token.

## Generation

- Stream: gen/now
- Promotion evidence: #8385 is blocked on a real native token source and current source inspection proves none reaches the wire.
- Demotion/retirement evidence: retire if an already-landed native receipt is discovered that carries exact ordered token indices and monotonic time.
- Invalidating assumption: the selected planner may not expose a pre-inference capability check without a narrower interface.
- Orthogonality: generation placement does not alter priority, lane leases, runtime defaults, or trace opt-in.

## Target operating envelope

- go toolchain: >= 1.26 version
- architecture: = 64 bits

## Witnessed operating envelope

- go toolchain: 1.26 version
- architecture: 64 bits

## Priority context

This is P1 because it unlocks a promotion gate and is deterministic/GPU-free. It does not supersede the active #8972 parity run or #9059 startup measurement.

## Gold-plating boundary

One native trace source and terminal receipt. Do not implement window aggregation, a general profiler, token content logging, or arbitrary provider trace inference.

## Work estimate

Estimate: 3 points. Uncertainty: gateway capability routing may need a narrow optional interface.

## Overall completion contribution

Contribution: 3/5 points toward #8385; the parent retains window aggregation and the live two-engine Mac campaign.

## Value

For: operators qualifying fak-native Qwen3.8 performance.
Problem: aggregate decode time cannot reveal position decay and content fragments are not token evidence.
Today: operators have whole-turn averages or ambiguous SSE chunks.
Better because: the native token-commit seam authors an opt-in ordered receipt that unsupported engines cannot forge.
Witness: one focused agent+gateway test binds native indices, timing, inert defaults, retry reset, and pre-inference refusal.

**Comment by anthony-chaudhary** [comment]
Coordinator readiness receipt (2026-08-26 PT): exact current `fak-dev@55db17f05` strict contract review reports `ok=true`, dispatchable S1, score 100, agent-context 100, strong independent witness; `fak dispatch issue-smallness-lint --issue 9071` is PASS with one deliverable and one witness. Source ownership remains the eight declared agent/gateway paths and no GPU is required. This is the next deterministic Mac-Qwen support leaf after the active #8972 hardware boundary, but no worker is started: guarded Codex preflight remains `REFUSE_AT_CAP` at live 8 / cap 6 / headroom -2, and the closed #9083 agent worktree must be reconciled as non-live before a fresh agent-lane collision check. No implementation or close claim is made.

**Comment by anthony-chaudhary** [comment]
Shipped on main in 39e9c9b124. The opt-in buffered fak-native path now emits a versioned, engine-authored token-indexed decode trace with monotonic commit timestamps and measured active compute time; unsupported, non-native, streaming, missing-trace, and default-off paths fail closed or remain unchanged. Independent evidence: isolated committed-tree tests for ./internal/agent and ./internal/gateway passed, focused receipt tests landed in the same commit, dos commit-audit is OK/diff-witnessed, and fak hooks pre-push plus fak sync push succeeded.
